File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/gift/LikeButton.java
package com.trtc.uikit.livekit.component.gift;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.FrameLayout;
import android.widget.ImageView;

import com.tencent.cloud.tuikit.engine.common.TUICommonDefine;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.tencent.qcloud.tuicore.TUILogin;
import com.trtc.uikit.livekit.R;
import com.trtc.uikit.livekit.common.ErrorLocalized;
import com.trtc.uikit.livekit.common.LiveKitLogger;
import com.trtc.uikit.livekit.component.gift.service.GiftService;
import com.trtc.uikit.livekit.component.gift.store.TUIGiftStore;
import com.trtc.uikit.livekit.component.gift.store.model.TUILikeData;

import java.util.Map;

@SuppressLint("ViewConstructor")
public class LikeButton extends FrameLayout {
    private static final LiveKitLogger LOGGER                = LiveKitLogger.getComponentLogger("LikeButton");
    private static final String        DEFAULT_AVATAR        = "https://liteav.sdk.qcloud.com/app/res/picture/voiceroom/avatar/user_avatar1.png";
    private static final int           LIKE_SEND_INTERVAL_MS = 6 * 1000;

    private int     mLikePendingCount   = 0;
    private long    mLikeLastSendTimeMS = 0;
    private boolean mHasPostLikeTask    = false;

    private final GiftService mGiftService  = new GiftService();
    private final Handler     mMainHandler  = new Handler(Looper.getMainLooper());
    private final Runnable    mSendLikeTask = () -> {
        LOGGER.info("mSendLikeTask run");
        mHasPostLikeTask = false;
        sendLike(mLikePendingCount, System.currentTimeMillis());
    };

    public LikeButton(Context context) {
        this(context, null);
    }

    public LikeButton(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LikeButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        LayoutInflater.from(context).inflate(R.layout.gift_layout_like_button, this);
        initView();
    }

    public void init(String roomId) {
        mGiftService.init(roomId, null);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        mGiftService.unInit();
    }

    private void initView() {
        setOnClickListener(v -> onLikeClick());
    }

    public void setImageResource(int resId) {
        ImageView imageLike = findViewById(R.id.iv_like);
        imageLike.setImageResource(resId);
    }

    public void onLikeClick() {
        String roomId = mGiftService.getRoomId();
        if (TextUtils.isEmpty(roomId)) {
            LOGGER.warn("roomId is null");
            return;
        }
        Long localLikeCount = TUIGiftStore.sharedInstance().mLocalLikeCountMap.get(roomId);
        localLikeCount = localLikeCount == null ? 1 : localLikeCount + 1;
        TUIGiftStore.sharedInstance().mLocalLikeCountMap.put(roomId, localLikeCount);
        mLikePendingCount++;
        long now = System.currentTimeMillis();
        long elapsed = now - mLikeLastSendTimeMS;
        TUIRoomDefine.UserInfo sender = createSelfUserInfo();
        Map<String, TUILikeData> map = TUIGiftStore.sharedInstance().mLikeDataMap.getValue();
        if (map != null) {
            map.put(roomId, new TUILikeData(sender));
            TUIGiftStore.sharedInstance().mLikeDataMap.setValue(map);
        }
        if (mLikeLastSendTimeMS == 0 || elapsed > LIKE_SEND_INTERVAL_MS) {
            mMainHandler.removeCallbacks(mSendLikeTask);
            mHasPostLikeTask = false;
            sendLike(mLikePendingCount, now);
        } else {
            if (!mHasPostLikeTask) {
                mMainHandler.postDelayed(mSendLikeTask, LIKE_SEND_INTERVAL_MS);
                mHasPostLikeTask = true;
            }
        }
    }

    private void sendLike(int count, long now) {
        mLikePendingCount = 0;
        mLikeLastSendTimeMS = now;
        mGiftService.sendLike(count, new TUIRoomDefine.ActionCallback() {
            @Override
            public void onSuccess() {

            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
                ErrorLocalized.onError(error);
                mLikePendingCount += count;
            }
        });
    }

    private TUIRoomDefine.UserInfo createSelfUserInfo() {
        TUIRoomDefine.UserInfo userInfo = new TUIRoomDefine.UserInfo();
        userInfo.userId = TextUtils.isEmpty(TUILogin.getUserId()) ? "" : TUILogin.getUserId();
        userInfo.userName = TextUtils.isEmpty(TUILogin.getNickName()) ? "" : TUILogin.getNickName();
        userInfo.avatarUrl = TextUtils.isEmpty(TUILogin.getFaceUrl()) ? DEFAULT_AVATAR : TUILogin.getFaceUrl();
        return userInfo;
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/gift/LikeButton.kt
package com.trtc.uikit.livekit.component.gift

import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import com.tencent.cloud.tuikit.engine.common.TUICommonDefine
import com.trtc.uikit.livekit.R
import com.trtc.uikit.livekit.common.ErrorLocalized.Companion.onError
import com.trtc.uikit.livekit.common.LiveKitLogger.Companion.getComponentLogger
import com.trtc.uikit.livekit.common.ui.BasicView
import io.trtc.tuikit.atomicxcore.api.CompletionHandler
import io.trtc.tuikit.atomicxcore.api.LikeStore

@SuppressLint("ViewConstructor")
class LikeButton @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
    BasicView(context, attrs, defStyleAttr) {
    private var likePendingCount = 0
    private var likeLastSendTimeMS: Long = 0
    private var hasPostLikeTask = false
    private var likeStore: LikeStore? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val sendLikeTask = Runnable {
        LOGGER.info("sendLikeTask run")
        hasPostLikeTask = false
        sendLike(likePendingCount, System.currentTimeMillis())
    }

    init {
        LayoutInflater.from(context).inflate(R.layout.gift_layout_like_button, this)
        initView()
    }

    override fun initStore() {
        likeStore = LikeStore.create(roomId)
    }

    override fun addObserver() {}

    override fun removeObserver() {}

    private fun initView() {
        setOnClickListener { v: View? -> onLikeClick() }
    }

    fun setImageResource(resId: Int) {
        val imageLike = findViewById<ImageView?>(R.id.iv_like)
        imageLike?.setImageResource(resId)
    }

    fun onLikeClick() {
        if (TextUtils.isEmpty(roomId)) {
            LOGGER.warn("roomId is null")
            return
        }
        likePendingCount++
        val now = System.currentTimeMillis()
        val elapsed = now - likeLastSendTimeMS
        if (likeLastSendTimeMS == 0L || elapsed > LIKE_SEND_INTERVAL_MS) {
            mainHandler.removeCallbacks(sendLikeTask)
            hasPostLikeTask = false
            sendLike(likePendingCount, now)
        } else {
            if (!hasPostLikeTask) {
                mainHandler.postDelayed(sendLikeTask, LIKE_SEND_INTERVAL_MS.toLong())
                hasPostLikeTask = true
            }
        }
    }

    private fun sendLike(count: Int, now: Long) {
        likePendingCount = 0
        likeLastSendTimeMS = now
        likeStore?.sendLike(count, object : CompletionHandler {
            override fun onSuccess() {}

            override fun onFailure(code: Int, desc: String) {
                onError(TUICommonDefine.Error.fromInt(code))
                likePendingCount += count
            }
        })
    }

    companion object {
        private val LOGGER = getComponentLogger("LikeButton")
        private const val LIKE_SEND_INTERVAL_MS = 6 * 1000
    }
}